CONTRIBUTING.md

```diff
@@ -0,0 +1,109 @@
+- Pull requests should remain focused in scope and avoid containing unrelated commits.
+- It is fine to have multiple, smaller commits as you work on the PR. We will let GitHub automatically squash your individual commits before merging.
+- *If you are adding new feature*:
+  - Please provide a convincing reason to add this feature. Ideally you should open a suggestion issue first and have it greenlighted before working on it.
+- *If you are fixing a bug*:
+  - If you are resolving a special issue, please add `(fix #xxxx[,#xxx])` (where `#xxxx` is the GitHub issue ID) to your pull request title, e.g. "`Update documentation (fix #9998)`".
+  - Provide a detailed description of the bug in the pull request. If possible, add a link to a live demo.
+
+### Committing Changes
+
+Possible TODO, see https://github.com/gatsbyjs/gatsby/issues/3920
+
+## Code Guidelines
+
+To ensure consistency throughout the source code, keep these guidelines in mind as you are working:
+
+- We use [Prettier](https://prettier.io/) to automate our code formatting
+- TODO semicolons, etc.?
+- TODO editorconfig?, see https://github.com/gatsbyjs/gatsby/issues/3922
+
+## What is Gatsby?
+
+Gatsby lets you build a blazing-fast a static PWA (Progressive Web App) with your data—whatever the source—using React.js, Webpack, GraphQL, modern JavaScript, CSS, and more. Our [home page](https://www.gatsbyjs.org/) provides a high-level introduction as well as detailed information that we are continuously trying to improve:
+
+* [Documentation and Guides](https://www.gatsbyjs.org/docs/)
+  * [What are Guide articles?](https://www.gatsbyjs.org/docs/gatsby-style-guide/#what-are-guide-articles)
+* [Tutorial](https://www.gatsbyjs.org/tutorial/)
+* [Features](https://www.gatsbyjs.org/features/)
+* [Blog](https://www.gatsbyjs.org/blog/)
+* [Showcase](https://github.com/gatsbyjs/gatsby#showcase)
+
+## Development Setup
+
+Gatsby is built using Node.js and supports versions back to v4 and npm to v3. You also need [Git](https://git-scm.com/) and the [Gatsby CLI](https://www.npmjs.com/package/gatsby-cli).
+
+Please refer to our [Get started](https://www.gatsbyjs.org/docs/) section for detailed instructions how to get install and use Gatsby's command line tool.
+
+After cloning this repository or installing a "Gatsby Starter", run the following command to install the project dependencies:
+
+```bash
+npm install # or yarn
+```
+
+## What is a "Gatsby Starter"?
+
+Gatsby uses “starters” for starting new projects. Starters are partially built Gatsby sites pre-configured to help you get moving faster. There are several official starters and many others contributed from the Gatsby community! [See the Starters page for the full list](https://www.gatsbyjs.org/docs/gatsby-starters/)
+
+### Commonly used NPM scripts
+
+- **`npm develop`**: TODO
+- **`npm build`**: TODO
+- **`npm format`**: TODO
+
+## Project Structure
```

Comment:
This is awesome. Several people have requested a list of typical folder structures for Gatsby projects.

Reply:
…but very incomplete right now. 😉

We definitively have this information present in the current documentation and the tutorial, e.g. at https://www.gatsbyjs.org/docs/building-with-components/, but maybe we can include an overview somewhere on gatsbyjs.org and link to that from CONTRIBUTING.md. This way we would only need to update that information in one place instead of having to update a bunch of CONTRIBUTING.md's.

I'm thinking (roughly) that we need to distinguish

* conventions (Gatsby expects folders to be named exactly like this and treats documents in there in a special way)
  * [`src/pages`](https://www.gatsbyjs.org/docs/building-with-components/#page-components)
  * [`src/layouts`](https://www.gatsbyjs.org/docs/building-with-components/#layout-components)
  * [`static`](https://www.gatsbyjs.org/docs/adding-images-fonts-files/#using-the-static-folder)
  * [`public`](https://www.gatsbyjs.org/tutorial/part-one/)
* recommendations (those can be named different, but we use those in our starters and naming them like "we do" makes sense)
  * `templates`
  * `utils`
  * `components`

…and will leave it at that—too late to communicate clearly. ;-)